### ISSUE TITLE
Robotics drift

### ISSUE THREAD

**Issue by gotmachine** [KSPModdingLibs/KSPCommunityFixes#13]
### Issue summary

When using a Breaking Grounds DLC robotic part, all child parts are affected by non-recoverable position/rotation drift.
Specifically, child parts of robotic parts have their "pristine" (as defined in the editor) coordinates updated to the current in-physics coordinates, which can differ significantly when the vessel is under stress due to forces like gravity, atmospheric drag, engine and RCS thrust, etc.
- This mean every time the game re-apply the saved "not-so-pristine-anymore" coordinates (ie, when timewarping or unloading the vessel), the original parts position/rotation are definitely lost and the vessel is permanently deformed. 
- The deformation gets worse over time, as every time the vessel goes in and out of physics, the in-physics forces apply a new deformation on top of initial one.

For the sake of clarity, two videos comparing the behavior of children parts of robotic parts and demonstrating the simplest way to reproduce that issue.

**Expected behavior when timewarping (no robotic part)**
The physics constraint is reverted and the parts "snap" back to their pristine positions :

https://user-images.githubusercontent.com/24925209/157562078-a0ae3863-a975-44bd-b248-da331132bb4e.mp4

**Robotics behavior when timewarping**
The physics constraint doesn't revert, and the resulting deformation is now permanent : 

https://user-images.githubusercontent.com/24925209/157561933-6b4eb293-53af-4500-9388-a2d6d70beb62.mp4

### Root cause

This is caused by `BaseServo.ApplyCoordsUpdate()` calling recursively the `Part.UpdateOrgPosAndRot()` method on all children parts. This method update the `Part.orgPos` and `Part.orgRot` fields (which are the "pristine" persisted part coordinates relative to the vessel root part) according the the current `Part.transform.position` and `Part.transform.rotation` values.

`BaseServo.ApplyCoordsUpdate()` is called continuously while a servo is moving, and on various occasions during the start / load / save events.

### Implemented fix

The current fix override the recursive call done in `BaseServo.ApplyCoordsUpdate()` to implement a custom logic. Instead of updating `Part.orgPos` and `Part.orgRot` based on the current in-physics part coordinates, it computes the servo induced translation/rotation offset, and apply that offset recursively to the current `Part.orgPos` and `Part.orgRot` values.

### Shortcomings and potential issues

- The patch currently doesn't cover the other coordinate update done in `BaseServo.RecurseAttachNodeUpdate()`. I haven't investigated as the current solution is satisfactory, and that code is probably fine as a result of the patch restoring pristine moving part coordinates prior to that method being called.
- The "pristine" coordinates we are computing might get slowly out of sync with the in-physics ones due to compound FP precision inconsistencies between our computations and the Unity ones. This being said, stress testing show part positions are visually stable after several hours of continuous servo operation.
- Computing the pristine coordinates adds some computational overhead when servos are in motion, but this is relatively insignificant compared to what those modules do already. 
- This is a behavior change that can have an adverse impact in some gameplay scenarios, especially while landed and under significant gravity. The stock "issue" also mean that all stress from gravity is "released" on a save/load or timewarp cycle. You get a deformed vessel, but it is now closer to being "at rest", which can help avoiding kraken events on scene load or when getting out of timewarp.

**Comment by ShadowGoat128134** [comment]
I have another potential solution. Just tell the game to remove all physics loads from any craft before it saves part positions. Less of a fix and more of removing the cause. I've been thinking of trying to make it myself, but I'm unsure how to go about making the mod.

**Comment by ShadowGoat128134** [comment]
My thinking with the second issue is that you’d need to give it a little bit to settle. Obviously it wouldnt go back to the exact same form it was in before launch, but you could at least imitate orbital conditions where drift is negligible.
Also the parts saving thing other than scene changes wouldn’t really matter would it? You’d really only need to worry about saving part positions when the vehicle is being unloaded/reloaded, because none of the intermediate saves actually update the craft until its unloaded.

**Comment by gotmachine** [comment]
> Also the parts saving thing other than scene changes wouldn’t really matter would it? 

Yes it would. From the top of my mind, there are at least 3 cases to handle (ie, cases where the current orgPos/orgRot will be restored later) :
- Scene changes
- Game saves (either manual or auto-saves)
- Vessel packing (every time a vessel enter non-physics timewarp, or when a vessel leaves physics loading range)

There is no "clean" way to delay those events to let physics run for few seconds. Even if you could do that without side issues (very unlikely), you would have to patch every possible entry point from all code paths in the stock codebase, and you would likely have tons of cases where other mods would bypass your hooks.

In conclusion, given that this isn't really a solution (you would need to delay all those actions by several seconds for the physics to settle vaguely reliably, with no guarantee that this actually work), and that it would be a huge mess to implement, I don't see the point of attempting this. Not to mention that when "re-enabling" physics again after doing your thing, you would inevitably trigger kraken events. This is just a terrible idea.

The "proper" solution I described would also require a lot of messing around and implementation work, but at least it would actually solve the issue.

**Comment by ShadowGoat128134** [comment]
Ah I see. Thanks for the explanation! I was kinda just trying to think of a
bandaid fix that wouldn't require completely reworking the craft saving
system. Too bad something half assed wouldn't work, since that would
simplify things a lot. I'm not sure why it didn't occur to me that time
warping was one of the instances where it restores the part positions.


*-Mitchell Woelfer*

**Comment by gotmachine** [comment]
For the sake of clarity, two videos comparing the behavior of children parts of robotic parts and demonstrating the simplest way to reproduce that issue.

### Expected behavior when timewarping
The physics constraint is reverted and the parts "snap" back to their pristine positions :

https://user-images.githubusercontent.com/24925209/157562078-a0ae3863-a975-44bd-b248-da331132bb4e.mp4

### Robotics behavior when timewarping
The physics constraint doesn't revert, and the resulting deformation is now permanent : 

https://user-images.githubusercontent.com/24925209/157561933-6b4eb293-53af-4500-9388-a2d6d70beb62.mp4

**Comment by gotmachine** [comment]
Fix implemented in KSPCF 1.9

**Comment by Lucaspec72** [comment]
Not 100% sure this isn't caused by another mod, but my console is flooded with NullReferenceExceptions
`
[EXC 10:33:34.877] NullReferenceException: Object reference not set to an instance of an object
	KSPCommunityFixes.BugFixes.RoboticsDrift+ServoInfo.PristineCoordsUpdate () (at <6d36e84722a243e9ba216b0d7c675f35>:0)
	KSPCommunityFixes.BugFixes.RoboticsDrift.BaseServo_RecurseCoordUpdate_Prefix (Expansions.Serenity.BaseServo __instance, Part p, UnityEngine.ConfigurableJoint ___servoJoint, UnityEngine.GameObject ___movingPartObject) (at <6d36e84722a243e9ba216b0d7c675f35>:0)
	(wrapper dynamic-method) Expansions.Serenity.BaseServo.Expansions.Serenity.BaseServo.RecurseCoordUpdate_Patch1(Expansions.Serenity.BaseServo,Part,Part)
	Expansions.Serenity.BaseServo.ApplyCoordsUpdate () (at <39c0323fb6b449a4aaf3465c00ed3c8d>:0)
	Expansions.Serenity.BaseServo.Update () (at <39c0323fb6b449a4aaf3465c00ed3c8d>:0)
	UnityEngine.DebugLogHandler:LogException(Exception, Object)
	ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
	UnityEngine.Debug:CallOverridenDebugHandler(Exception, Object)`

**Comment by JonnyOThan** [comment]
NOTE: I have KSPCF 1.12.2 installed.

I'm not sure if this is related or should be treated as its own issue.  I have a rover attached to the bottom of my lander with a piston and a klaw.  I saved the game with the piston partway extended so that the rover's wheels were touching the ground.  When I loaded it, the klaw's relative position to the end of the piston was no longer correct so the rover was in the wrong position.  It seems like on every save/load it gets further away.
![aa_none](https://user-images.githubusercontent.com/9804078/168445492-f95ee5d1-712a-4a97-beec-8f0edcb5e44c.png)
[persistent.txt](https://github.com/KSPModdingLibs/KSPCommunityFixes/files/8693725/persistent.txt)

**Comment by gotmachine** [comment]
@JonnyOThan This is likely the same issue as [the one someone just reported on the forums](https://forum.kerbalspaceprogram.com/index.php?/topic/204002-18-112-kspcommunityfixes-bugfixes-and-qol-tweaks/&do=findComment&comment=4132262).

I changed something in KSPCF 1.12.2 to fix another bug in the RoboticsDrift patch involving rotation servos, but I forgot to check if that was working correctly with translation servos. Should hopefully be an easy fix. Sorry for the inconvenience.

**Comment by gotmachine** [comment]
Should be fixed in release 1.13.2

However, this won't correct things for existing vessels, sorry for the inconvenience...

**Comment by JonnyOThan** [comment]
Well, glad it's fixed!  I think I don't even need to pick the rover up again on this mission, so no inconvenience at all.
